util/pingpong.c

```diff
@@ -0,0 +1,2111 @@
+	int sizes_enabled;
+	int options;
+	int argc;
+	char **argv;
+};
+
+#define ADDR_OPTS "b:p:s:"
+#define INFO_OPTS "n:f:e:"
+#define CS_OPTS ADDR_OPTS "I:S:"
+#define BENCHMARK_OPTS "v"
+
+#define INIT_OPTS (struct ft_opts) \
+	{	.options = 0, \
+		.iterations = 1000, \
+		.transfer_size = 1024, \
+		.sizes_enabled = FT_DEFAULT_SIZE, \
+		.argc = argc, .argv = argv \
+	}
+
+#define FT_STR_LEN 32
+#define FT_MAX_CTRL_MSG 64
+#define FT_MR_KEY 0xC0DE
+#define FT_MSG_MR_ACCESS (FI_SEND | FI_RECV)
+#define FT_RMA_MR_ACCESS (FI_READ | FI_WRITE | FI_REMOTE_READ | FI_REMOTE_WRITE)
+
+#define FT_PRINTERR(call, retv) \
+	do { fprintf(stderr, call "(): %s:%d, ret=%d (%s)\n", __FILE__, __LINE__, \
+			(int) retv, fi_strerror((int) -retv)); } while (0)
+
+#define FT_LOG(level, fmt, ...) \
+	do { fprintf(stderr, "[%s] %s:%d: " fmt "\n", level, __FILE__, \
+			__LINE__, ##__VA_ARGS__); } while (0)
+
+#define FT_LOG_(level, fmt, ...) \
+	do { fprintf(stderr, "[%s] %s:%d: " fmt, level, __FILE__, \
+			__LINE__, ##__VA_ARGS__); } while (0)
+
+#define FT_ERR(fmt, ...) FT_LOG("error", fmt, ##__VA_ARGS__)
+#define FT_WARN(fmt, ...) FT_LOG("warn", fmt, ##__VA_ARGS__)
+
+int _activate_debug = 0;
+#define FT_ACTIVATE_DEBUG _activate_debug
+
+#define FT_DEBUG(fmt, ...)						\
+	do {								\
+		if (_activate_debug) {					\
+			FT_LOG_("debug", fmt, ##__VA_ARGS__);		\
+		}							\
+	} while (0)
+
+#define FT_EQ_ERR(eq, entry, buf, len) \
+	FT_ERR("eq_readerr: %s", fi_eq_strerror(eq, entry.prov_errno, \
+				entry.err_data, buf, len))
+
+#define FT_CQ_ERR(cq, entry, buf, len) \
+	FT_ERR("cq_readerr: %s", fi_cq_strerror(cq, entry.prov_errno, \
+				entry.err_data, buf, len))
+
+#define FT_CLOSE_FID(fd)						\
+	do {								\
+		int ret;						\
+		if ((fd)) {						\
+			ret = fi_close(&(fd)->fid);			\
+			if (ret)					\
+				FT_ERR("fi_close (%d) fid %d",		\
+					ret, (int) (fd)->fid.fclass);	\
+			fd = NULL;					\
+		}							\
+	} while (0)
+
+#define FT_CLOSEV_FID(fd, cnt)			\
+	do {					\
+		int i;				\
+		if (!(fd))			\
+			break;			\
+		for (i = 0; i < (cnt); i++) {	\
+			FT_CLOSE_FID((fd)[i]);	\
+		}				\
+	} while (0)
+
+
+#define FT_PROCESS_QUEUE_ERR(readerr, rd, queue, fn, str)	\
+	do {							\
+		if (rd == -FI_EAVAIL) {				\
+			readerr(queue, fn " " str);		\
+		} else {					\
+			FT_PRINTERR(fn, rd);			\
+		}						\
+	} while (0)
+
+#define FT_PROCESS_EQ_ERR(rd, eq, fn, str) \
+	FT_PROCESS_QUEUE_ERR(eq_readerr, rd, eq, fn, str)
+
+#define FT_PRINT_OPTS_USAGE(opt, desc) fprintf(stderr, " %-20s %s\n", opt, desc)
+
+#define MIN(a,b) (((a)<(b))?(a):(b))
+#define MAX(a,b) (((a)>(b))?(a):(b))
+#define ARRAY_SIZE(A) (sizeof(A)/sizeof(*A))
+
+#define TEST_ENUM_SET_N_RETURN(str, enum_val, type, data)	\
+	TEST_SET_N_RETURN(str, #enum_val, enum_val, type, data)
+
+#define TEST_SET_N_RETURN(str, val_str, val, type, data)	\
+	do {							\
+		if (!strncmp(str, val_str, strlen(val_str))) {	\
+			*(type *)(data) = val;			\
+			return 0;				\
+		}						\
+	} while (0)
```

Comment:
The `TEST_ENUM_*` macros are never used.

I'll stop making these repetitive comments, but please:

* Remove all macros that are not used
* Remove all macros that are only used once